src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.cs

```diff
@@ -98,22 +98,13 @@ internal sealed class ContainedDocument : ForegroundThreadAffinitizedObject, IVi
             _optionService = _workspace.Services.GetService<IOptionService>();
             _hostType = GetHostType();
 
-            var rdt = (IVsRunningDocumentTable)componentModel.GetService<SVsServiceProvider>().GetService(typeof(SVsRunningDocumentTable));
-
-            IVsHierarchy sharedHierarchy;
-            uint itemIdInSharedHierarchy;
-            var isSharedHierarchy = LinkedFileUtilities.TryGetSharedHierarchyAndItemId(hierarchy, itemId, out sharedHierarchy, out itemIdInSharedHierarchy);
-
-            var filePath = isSharedHierarchy
-                ? rdt.GetMonikerForHierarchyAndItemId(sharedHierarchy, itemIdInSharedHierarchy)
-                : rdt.GetMonikerForHierarchyAndItemId(hierarchy, itemId);
-
-            // we couldn't look up the document moniker in RDT for a hierarchy/item pair
-            // Since we only use this moniker as a key, we could fall back to something else, like the document name.
-            if (filePath == null)
+            string filePath;
+            if (!ErrorHandler.Succeeded(((IVsProject)hierarchy).GetMkDocument(itemId, out filePath)))
             {
-                Debug.Assert(false, "Could not get the document moniker for an item in its hierarchy.");
-                filePath = hierarchy.GetDocumentNameForHierarchyAndItemId(itemId);
+                // we couldn't look up the document moniker from an hierarchy for an itemid.
+                // Since we only use this moniker as a key, we could fall back to something else, like the document name.
+                Debug.Assert(false, "Could not get the document moniker for an item from its hierarchy.");
+                hierarchy.TryGetItemName(itemId, out filePath);
```

Comment:
If it fails, then filePath will come back null and the DocumentKey creation below will throw. So, it will explode very quickly, not "much later". However, I'm with @rchande - I'd rather assert here.